pkg/envoy/cds/builder.go

```diff
@@ -28,9 +31,139 @@ import (
+	b.inboundMeshTrafficPolicy = inboundTrafficPolicy
+	return b
+}
+
+func (b *clusterBuilder) SetEgressTrafficPolicy(egressTrafficPolicy *trafficpolicy.EgressTrafficPolicy) *clusterBuilder {
+	b.egressTrafficPolicy = egressTrafficPolicy
+	return b
+}
+
+func (b *clusterBuilder) SetEnvoyOutBoundPassthroughCluster(envoyOutboundPassthroughCluster string) *clusterBuilder {
+	b.envoyOutboundPassthroughCluster = envoyOutboundPassthroughCluster
+	return b
+}
+
+func (b *clusterBuilder) SetEgressEnabled(egressEnabled bool) *clusterBuilder {
+	b.egressEnabled = egressEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetMetricsEnabled(metricsEnabled bool) *clusterBuilder {
+	b.metricsEnabled = metricsEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetTracingEnabled(tracingEnabled bool) *clusterBuilder {
+	b.tracingEnabled = tracingEnabled
+	return b
+}
+
+func (b *clusterBuilder) SetTracingAddress(tracingAddress string) *clusterBuilder {
+	b.tracingAddress = tracingAddress
+	return b
+}
+
+func (b *clusterBuilder) SetTracingPort(tracingPort uint32) *clusterBuilder {
+	b.tracingPort = tracingPort
+	return b
+}
+
 // replacer used to configure an Envoy cluster's altStatName
 var replacer = strings.NewReplacer(".", "_", ":", "_")
 
+func (b *clusterBuilder) Build() ([]*xds_cluster.Cluster, error) {
+	var clusters []*xds_cluster.Cluster
+
+	// Build upstream clusters based on allowed outbound traffic policies
+	if b.outboundMeshTrafficPolicy != nil {
+		clusters = append(clusters, upstreamClustersFromClusterConfigs(b.proxyIdentity, b.outboundMeshTrafficPolicy.ClustersConfigs, b.sidecarSpec)...)
+	}
+
+	// Build local clusters based on allowed inbound traffic policies
+	if b.inboundMeshTrafficPolicy != nil {
+		clusters = append(clusters, localClustersFromClusterConfigs(b.inboundMeshTrafficPolicy.ClustersConfigs)...)
+	}
+
+	// Add egress clusters based on applied policies
+	if b.egressTrafficPolicy != nil {
+		clusters = append(clusters, getEgressClusters(b.egressTrafficPolicy.ClustersConfigs)...)
+	}
+
+	// Add an outbound passthrough cluster for egress if global mesh-wide Egress is enabled
+	if b.egressEnabled {
+		outboundPassthroughCluster, err := getOriginalDestinationEgressCluster(b.envoyOutboundPassthroughCluster, nil)
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.ErrGettingOrgDstEgressCluster.String()).
+				Msgf("Failed to passthrough cluster for egress for proxy %s", b.envoyOutboundPassthroughCluster)
+			return nil, err
+		}
+		clusters = append(clusters, outboundPassthroughCluster)
+	}
+
+	// Add an inbound prometheus cluster (from Prometheus to localhost)
+	if b.metricsEnabled {
+		clusters = append(clusters, getPrometheusCluster())
+	}
+
+	// Add an outbound tracing cluster (from localhost to tracing sink)
+	if b.tracingEnabled {
+		clusters = append(clusters, getTracingCluster(b.tracingAddress, b.tracingPort))
+	}
+
+	return removeDups(clusters), nil
```

Comment:
I don't think we need remove dups anymore, that was an edge case from the original multicluster attempt

Reply:
seems like this other test case also expects duplicates to exist if an egress traffic cluster config lists the same cluster twice, is this also multicluster related?
https://github.com/openservicemesh/osm/blob/main/pkg/envoy/cds/response_test.go#L437-L460